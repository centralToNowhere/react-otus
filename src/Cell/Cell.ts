export interface ICell {
  x: number;
  y: number;
}

export type IndexedCells = Array<1 | 0>;

export const maxCellsAmount = 20000;
