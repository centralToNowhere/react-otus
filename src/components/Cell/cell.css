.cell {
    outline: 1px solid white;
    text-align: center;
    width: 10px;
    height: 10px;
}