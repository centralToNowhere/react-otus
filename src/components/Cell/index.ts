export { Cell } from "./Cell";
export type { ICell, ICellProps } from "./Cell";
