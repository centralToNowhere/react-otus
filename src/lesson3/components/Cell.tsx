import React from "react";
import { useSelector } from "react-redux";
import { IState } from "../redux/store";

export interface ICell {
  number: number;
  backgroundColor: string;
  alive: boolean;
}

export interface ICellProps {
  number: number;
  width: number;
  onCellClick: (number: number) => void;
}

export function isCell(cell: ICell | unknown): cell is ICell {
  if (cell === null || typeof cell !== "object") {
    return false;
  }

  return (
    "number" in (cell as ICell) &&
    "backgroundColor" in (cell as ICell) &&
    "alive" in (cell as ICell)
  );
}

const Cell = (props: ICellProps) => {
  const { number, onCellClick, width } = props;
  const cell: ICell | undefined = useSelector((state: IState) =>
    state.cells.find((c) => {
      return c.number === number;
    })
  );

  if (!isCell(cell)) {
    return null;
  }

  const cellContainerStyles = {
    width,
    height: width,
    fontSize: width,
  };

  const cellStyles: object = {
    backgroundColor: cell.alive ? cell.backgroundColor : null,
  };

  return (
    <div
      className="container-cell"
      style={cellContainerStyles}
      onClick={() => {
        onCellClick(number);
      }}
    >
      <div className="cell" style={cellStyles} data-testid="cell">
        {cell.alive ? number : null}
      </div>
    </div>
  );
};

export default Cell;
