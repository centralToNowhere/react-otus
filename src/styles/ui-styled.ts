export const COLORS = {
  primary: "#9a949a",
  secondary: "#e8d5f0",
  accent: "#ff215c",
  activeCellBg: "#9a949a",
  border: "rgb(118, 118, 118)",
};

export const BREAKPOINTS = {
  mobileStart: "320px",
  mobileEnd: "480px",
  sm: "576px",
  md: "768px",
  lg: "992px",
  xl: "1200px",
};
