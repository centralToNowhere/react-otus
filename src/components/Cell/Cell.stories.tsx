import React from "react";
import { Story, Meta } from "@storybook/react";
import { Cell, ICellProps } from "./Cell";

export default {
  title: "Components/Cell",
  component: Cell,
  args: {
    cssClassName: "cell",
    isActive: true,
    number: 1,
  },
} as Meta;

const Template: Story<ICellProps> = (args) => <Cell {...args} />;

export const CellActive = Template;
