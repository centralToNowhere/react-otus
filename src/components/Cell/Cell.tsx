import React from "react";
import "@/components/Cell/cell.css";
import styled from "@emotion/styled";
import { COLORS } from "@/styles/ui-styled";

export interface ICell {
  x: number;
  y: number;
}

export interface ICellProps {
  isActive: boolean;
  number: number;
  cssClassName: string;
}

export const minCellSize = 10;

export const isCell = (cell: ICell): cell is ICell => {
  return (
    cell !== null &&
    typeof cell === "object" &&
    typeof cell.x === "number" &&
    typeof cell.y === "number"
  );
};

export class Cell extends React.PureComponent<ICellProps> {
  constructor(props: ICellProps) {
    super(props);
  }

  render() {
    return (
      <StyledCell
        className={this.props.cssClassName}
        data-testid="cell"
        data-state={this.props.isActive}
        aria-label={String(this.props.number)}
      />
    );
  }
}

const StyledCell = styled.div`
  border: 1px solid ${COLORS.activeCellBg};
  text-align: center;
  width: 10px;
  height: 10px;
`;
