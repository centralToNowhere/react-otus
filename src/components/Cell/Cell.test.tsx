import React from "react";
import { render, screen } from "@testing-library/react";
import { Cell } from "./Cell";
import { COLORS } from "@/styles/ui-styled";

describe("Cell tests", () => {
  it("should render cell", () => {
    const { asFragment } = render(
      <Cell size={30} isActive={true} number={1} />
    );

    const cell = screen.getByTestId("cell");

    expect(cell).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render active cell", () => {
    render(<Cell size={30} isActive={true} number={1} />);

    const cell = screen.getByTestId("cell");

    expect(cell).toHaveStyle({
      background: COLORS.activeCellBg,
    });
  });

  it("should render inactive cell", () => {
    render(<Cell size={30} isActive={false} number={1} />);

    const cell = screen.getByTestId("cell");

    expect(cell).not.toHaveStyle({
      background: COLORS.activeCellBg,
    });
  });
});
