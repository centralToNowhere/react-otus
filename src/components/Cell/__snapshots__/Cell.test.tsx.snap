// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell tests should render cell 1`] = `
<DocumentFragment>
  .emotion-0 {
  border: 1px solid #9a949a;
  text-align: center;
  width: 10px;
  height: 10px;
}

<div
    aria-label="1"
    class="cell cell-active emotion-0 emotion-1"
    data-state="true"
    data-testid="cell"
  />
</DocumentFragment>
`;
